Sources/Utility/Process.swift

```diff
@@ -0,0 +1,295 @@
+        let devNull = strdup("/dev/null")
+        defer { free(devNull) }
+        // Open /dev/null as stdin.
+        posix_spawn_file_actions_addopen(&fileActions, 0, devNull, O_RDONLY, 0)
+
+        var rv = libc.pipe(&saveOutputPipe)
+        guard rv == 0 else {
+            throw SystemError.pipe(rv)
+        }
+
+        if outputStream != nil {
+            // Open the write end of the pipe as stdout and stderr, if desired.
+            posix_spawn_file_actions_adddup2(&fileActions, saveOutputPipe[1], 1)
+            posix_spawn_file_actions_adddup2(&fileActions, saveOutputPipe[1], 2)
+
+            // Close the other ends of the pipe.
+            posix_spawn_file_actions_addclose(&fileActions, saveOutputPipe[0])
+            posix_spawn_file_actions_addclose(&fileActions, saveOutputPipe[1])
+        } else {
+            posix_spawn_file_actions_adddup2(&fileActions, 1, 1)
+            posix_spawn_file_actions_adddup2(&fileActions, 2, 2)
+        }
+
+        let argv: [UnsafeMutablePointer<CChar>?] = arguments.map{ $0.withCString(strdup) }
+        defer { for case let arg? in argv { free(arg) } }
+        
+        let env: [UnsafeMutablePointer<CChar>?] = environment.map{ "\($0.0)=\($0.1)".withCString(strdup) }
+        defer { for case let arg? in env { free(arg) } }
+
+        rv = posix_spawnp(&processID, argv[0], &fileActions, &attributes, argv + [nil], env + [nil])
+
+        posix_spawn_file_actions_destroy(&fileActions)
+        posix_spawnattr_destroy(&attributes)
+
+        guard rv == 0 else {
+            throw SystemError.posix_spawn(rv, arguments)
+        }
+    }
+
+    /// Blocks the calling process until the subprocess finishes execution.
+    ///
+    /// - Throws: Process.Error
+    public func waitUntilExit() throws {
+        assert(launched, "The process is not yet launched.")
+        // FIXME: We should probably read the output from somewhere else, maybe in a seperate thread?
+        if let outputStream = outputStream {
+            try readOutput(on: outputStream)
+        }
+        var exitStatus: Int32 = 0
+        var result = waitpid(processID, &exitStatus, 0)
+        while (result == -1 && errno == EINTR) {
+            result = waitpid(processID, &exitStatus, 0)
+        }
+        if result == -1 {
+            throw SystemError.waitpid(errno)
+        }
+
+        if exitStatus & 0x7f == 0 {
+            self.returnCode = (exitStatus >> 8) & 0xff
+        } else {
+            // We received a signal.
+            // FIXME: We may need a better way to represent this.
+            signalled = true
+            self.returnCode = exitStatus
+        }
+    }
+
+    /// Reads the output from save output pipes.
+    ///
+    /// - Throws: Process.Error
+    private func readOutput(on stream: OutputByteStream) throws {
+        assert(saveOutputPipe != [0, 0], "Pipes not initialized.")
+        // Close the write end of the output pipe.
+        let rv = close(saveOutputPipe[1])
```

Comment:
This close should be done immediately after spawning the process, and then discarded.